PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+                    combined['Overall_lineage'][isolate] = combined['Overall_lineage'][isolate] + '-' + lineage_string
+                else:
+                    combined['Overall_lineage'][isolate] = lineage_string
+            lFile.write(',' + combined['Overall_lineage'][isolate])
+            if qlist is not None:
+                if isolate in qlist:
+                    lFile.write(',Added')
+                else:
+                    lFile.write(',Existing')
+            lFile.write('\n')
+
+    return combined
+
+def run_clustering_for_R(R, null_cluster_value = None, qlist = None, existing_scheme = False, distances_length = None, distances_type = None):
+    """ Clusters isolates into lineages based on their
+    relative distances using a single R to enable
+    parallelisation.
+
+    Args:
+        R (int)
+            Integer specifying the maximum rank of neighbour used
+            for clustering. Should be changed to int list for hierarchical
+            clustering.
+        null_cluster_value (int)
+            Used to denote as-yet-unclustered isolates.
+        qlist (list)
+            List of query sequences being added to an existing clustering.
+            Should be included within rlist.
+        use_existing (bool)
+            Whether to extend a previously generated analysis or not.
+            
+    Returns:
+        lineage_clustering (dict)
+            Assignment of each isolate to a cluster.
+        lineage_seed (dict)
+            Seed isolate used to initiate each cluster.
+        neighbours (nested dict)
+            Neighbour relationships between isolates for R.
+    """
+    
+    # load shared memory objects
+    existing_distance_shm = shared_memory.SharedMemory(name = 'shm_distances')
+    distances = np.ndarray(distances_length, dtype = distances_type, buffer = existing_distance_shm.buf)
+    existing_rank_shm = shared_memory.SharedMemory(name = 'shm_distance_ranks')
+    distance_ranks = np.ndarray(distances_length, dtype = 'int64', buffer = existing_rank_shm.buf)
+    isolate_list = shared_memory.ShareableList(name = 'shm_isolate_list')
+    
+    # calculate row labels
+    # this is inefficient but there appears to be no way of sharing
+    # strings between processes efficiently
+    row_labels = list(iter(iterDistRows(isolate_list, isolate_list, self = True)))
+    # reorder by sorted distances
+    row_labels = [row_labels[i] for i in distance_ranks]
+    
+    lineage_clustering = {i:null_cluster_value for i in isolate_list}
+    previous_lineage_clustering = lineage_clustering
+    lineage_seed = {}
+    neighbours = {}
+    
+    if existing_scheme is not None:
+        with open(existing_scheme, 'rb') as pickle_file:
+            lineage_clustering_overall, lineage_seed_overall, neighbours_overall, rank_list = pickle.load(pickle_file)
+        # focus on relevant data
+        lineage_clustering = lineage_clustering_overall[R]
```

Comment:
Is R an int or str (I guess str, they look like dicts above)

Reply:
Specified as rank (int) in 01cd435a731ae8cbc268fe7eacc4ea0dc7f42eb9.